group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -119,6 +131,18 @@ public String toString() {
      */
     private final TimelineHashMap<Uuid, TimelineHashMap<Integer, Integer>> currentPartitionEpoch;
 
+    /**
+     * The next metadata refresh time. It consists of a timestamp in milliseconds together with
+     * the group epoch at the time of setting it. The metadata refresh time is considered as a
+     * soft state (read that it is not stored in a timeline data structure). It is like this
+     * because it is not persisted to the log. The group epoch is here to ensure that the
+     * next metadata refresh time is invalidated if the group epoch does not correspond to
+     * the current group epoch. This can happen if the next metadata refresh time is updated
+     * after having refreshed the metadata but the write operation failed. In this case, the
+     * time is not automatically rollback.
```

Comment:
nit: automatically rolled back.

Also in this case, do we not update the refresh time until the epoch bump and write succeed? And we will keep refreshing the metadata in the meantime?

Reply:
It is actually the other way around. The refresh time is updated immediately but it is not rolled back if the write failed. This is the reason why I have included the group epoch.